### PR TITLE
misc: ensure 'its' function type excludes null and undefined (#28872)

### LINKED ISSUE #28872: `its` type should exclude `null` and `undefined`
### Current behavior

A Cypress test that relies on `its` to yield a value that is not `null` or `undefined` will be flagged as an error in TypeScript, even though the [docs assert](https://docs.cypress.io/api/commands/its#Assertions) that "`.its()` will automatically retry until it has a property that is not `null` or `undefined`."

### Desired behavior

Tests shouldn't have to handle values that can't occur.

### Test code to reproduce

```ts
describe('contrived example', () => {
  specify('log frameElement', () => {
    cy.window()
      .its('frameElement')
      .then(frameElement => {
        cy.log(frameElement.id)
      })
  })
})
```

This test runs (and passes, for what it's worth). But TypeScript reports:

```
cypress/e2e/scenarios/contrived.cy.ts:6:16 - error TS18047: 'frameElement' is possibly 'null'.

6         cy.log(frameElement.id)
                 ~~~~~~~~~~~~
```

In this example, I can work around it by logging `frameElement?.id || ''`, but it would be better not to have to. And although workarounds are probably always possible, they're not always that simple.

(My actual code involves a third-party analytics library that adds a property to `window` with an object I want to stub, but the property isn't available immediately on page load, so I have to wait for it. In JavaScript, `its` is great for that! But I'm converting the tests to TypeScript, and TypeScript doesn't know it's great in that way.)

### Cypress Version

13.6.4

### Node version

18.17.0

### Operating System

MacOS 14.2.1

### Debug Logs

_No response_

### Other

It looks to me like this can be fixed by making the resulting `Chainable` subject non-nullable – changing:

```ts
    its<K extends keyof Subject>(propertyName: K, options?: Partial<Loggable & Timeoutable>): Chainable<Subject[K]>
```

to:

```ts
    its<K extends keyof Subject>(propertyName: K, options?: Partial<Loggable & Timeoutable>): Chainable<NonNullable<Subject[K]>>
```

I did that as a local override, and my TypeScript errors went away. That may or may not be the best solution.

### PR DESCRIPTION
**Pull request by AkshatHotCode**
This fix addresses an issue where the 'its' function could return null or undefined, causing unexpected behavior in certain scenarios. The change introduces a more robust type check to ensure that the returned value is always non-null and non-undefined.

- Closes https://github.com/cypress-io/cypress/issues/28872

### Additional details
The change was necessary to address an issue where the 'its' function could potentially return null or undefined. This behavior led to unexpected outcomes in certain scenarios. The fix ensures that the 'its' function consistently returns a value that is non-null and non-undefined.
The affected area is the 'its' function, specifically its type signature. Prior to the fix, the function had the potential to return null or undefined. This change impacts any usage of the 'its' function, ensuring that users can rely on a non-null and non-undefined result.
The implementation involves a modification to the 'its' function to include a more robust type check. This check ensures that the returned value is guaranteed to be non-null and non-undefined. 

### PR Tasks

- [na] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]
@AkshatHotCode There are some errors in the lint types and unit tests. Could you look at those from the failing CircleCI checks above? We'll also need a changelog entry for this change. https://github.com/cypress-io/cypress/blob/develop/guides/writing-the-cypress-changelog.md

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]
@AkshatHotCode Looks like there's some Lint errors - likely from the formatting changes you made that were unrelated to the actual fix. https://app.circleci.com/pipelines/github/cypress-io/cypress/60042/workflows/b5c2b3d3-bbae-40df-92db-f7058dd2a131/jobs/2498299

Screenshot 2024-02-27 at 9 59 43 AM

**Comment by jennifer-shehane** [comment]
@AkshatHotCode Will you have time to address the lint errors? We'll have to close this PR otherwise.

**Comment by AkshatHotCode** [comment]
I will address this.

**Comment by AkshatHotCode** [comment]
@jennifer-shehane Changes are being made, apologies for the delay.

File: cli/types/cypress.d.ts [hunk 1]
```diff
@@ -592,14 +592,13 @@ declare namespace Cypress {
       /**
        * Add a custom parent command
        * @see https://on.cypress.io/api/commands#Parent-Commands
-       */
-      add<T extends keyof Chainable>(name: T, options: CommandOptions & {prevSubject: false}, fn: CommandFn<T>): void
+       */add<T extends keyof Chainable>(name: T, options: CommandOptions & { prevSubject: false }, fn: CommandFn<T>): void
```
Comment (ryanthemanuel) [inline cli/types/cypress.d.ts]:
```suggestion
       */
       add<T extends keyof Chainable>(name: T, options: CommandOptions & { prevSubject: false }, fn: CommandFn<T>): void
```

File: cli/types/cypress.d.ts [hunk 2]
```diff
@@ -593,13 +593,13 @@ declare namespace Cypress {
        * Add a custom parent command
        * @see https://on.cypress.io/api/commands#Parent-Commands
        */
-      add<T extends keyof Chainable>(name: T, options: CommandOptions & {prevSubject: false}, fn: CommandFn<T>): void
+       add<T extends keyof Chainable>(name: T, options: CommandOptions & { prevSubject: false }, fn: CommandFn<T>): void
```
Comment (ryanthemanuel) [inline cli/types/cypress.d.ts]:
```suggestion
      add<T extends keyof Chainable>(name: T, options: CommandOptions & { prevSubject: false }, fn: CommandFn<T>): void
```

**Review by ryanthemanuel — APPROVED** [review]

**Comment by jennifer-shehane** [comment]
@AkshatHotCode Looking back at this, sorry for the delay

**Review by jennifer-shehane — APPROVED** [review]

**Pull request merged**